craft_parts/plugins/cargo_package_plugin.py

```diff
@@ -0,0 +1,163 @@
+
+    plugin: Literal["cargo-package"] = "cargo-package"
+
+    # part properties required by the plugin
+    cargo_package_features: UniqueList[str] = pydantic.Field(default_factory=list)
+    cargo_package_cargo_command: str = "cargo"
+    source: str  # pyright: ignore[reportGeneralTypeIssues]
+    after: UniqueList[str] | None = None
+
+
+class CargoPackagePluginEnvironmentValidator(validator.PluginEnvironmentValidator):
+    """Validate the environment for the cargo package plugin."""
+
+    @override
+    def validate_environment(
+        self, *, part_dependencies: list[str] | None = None
+    ) -> None:
+        """Ensure the environment has the dependencies to build crates.
+
+        :param part_dependencies: A list of the parts this part depends on.
+        """
+        options = cast(CargoPackagePluginProperties, self._options)
+        if "rust-deps" in (part_dependencies or []):
+            # If we have a rust-deps dependency, we don't need to check for cargo.
+            return
+
+        self.validate_dependency(
+            dependency=options.cargo_package_cargo_command,
+            plugin_name=options.plugin,
+            part_dependencies=part_dependencies,
+        )
+
+
+class CargoPackagePlugin(Plugin):
+    """Build a Cargo crate and publish it to a local directory registry."""
+
+    properties_class = CargoPackagePluginProperties
+    validator_class = CargoPackagePluginEnvironmentValidator
+
+    @override
+    def get_build_snaps(self) -> set[str]:
+        """Return a set of required snaps to install in the build environment."""
+        return set()
+
+    @override
+    def get_build_packages(self) -> set[str]:
+        """Return a set of required packages to install in the build environment."""
+        return set()
+
+    @property
+    def _registry_dir(self) -> pathlib.Path:
+        return self._part_info.project_info.dirs.stage_dir / "cargo_registry"
+
+    @property
+    def _registry_output_dir(self) -> pathlib.Path:
+        return self._part_info.part_install_dir / "cargo_registry"
+
+    @override
+    def get_build_environment(self) -> dict[str, str]:
+        """Return a dictionary with the environment to use in the build step."""
+        return {}
+
+    def _get_package_command(self) -> str:
+        options = cast(CargoPackagePluginProperties, self._options)
+        cargo = options.cargo_package_cargo_command
+        features = ",".join(options.cargo_package_features)
+        if features:
+            features = f"--features {features}"
+        return f"{cargo} package --no-verify {features}"
+
+    @override
+    def get_build_commands(self) -> list[str]:
+        """Return a list of commands to run during the build step."""
+        # We do not want this implementation detail exposed in the run script
+        # This sets the default registry for all parts to use.
+        cargo_config = pathlib.Path("~/.cargo/config.toml").expanduser().resolve()
+        if not cargo_config.exists():
```

Comment:
how does this play with destructive mode?

Reply:
It will destroy your user's cargo config in destructive mode.

Reply:
With ST143 though we'll instead set `CARGO_HOME` and use that.

Reply:
This now uses a project-specific CARGO_HOME under the work directory.